File: examples/demo-nodejs/src/demo-llm.ts
import OpenAI from 'openai'
import chalk from 'chalk'
import { setGlobalDispatcher, ProxyAgent } from 'undici'

type SunraChatCompletionParams = OpenAI.Chat.Completions.ChatCompletionCreateParamsStreaming & {
  provider?: {
    only?: string[];
  };
};

const apiKey = process.env.SUNRA_KEY || ''
if (!apiKey) {
  console.error(chalk.red('SUNRA_KEY is not set'))
  console.error(chalk.yellow('Get your API key from https://sunra.ai/dashboard/api-tokens'))
  process.exit(1)
}

/**
 * Demo script for LLM endpoint using OpenAI SDK directly.
 * This demonstrates how to use Sunra's LLM API with the standard OpenAI client.
 */
const main = async () => {
  try {
    console.log(chalk.green('Initializing OpenAI client with Sunra endpoint...'))

    // Configure undici (which fetch uses) to respect proxy settings
    const proxyUrl = process.env.HTTPS_PROXY || process.env.HTTP_PROXY
    if (proxyUrl) {
      console.log(chalk.yellow(`Using proxy: ${proxyUrl}`))
      setGlobalDispatcher(new ProxyAgent(proxyUrl))
    }

    // Initialize the standard OpenAI client with Sunra's compatible endpoint.
    const client = new OpenAI({
      apiKey,
      baseURL: 'https://api-llm.sunra.ai/v1',
    })

    console.log(chalk.green('Sending streaming chat completion request...'))

    const request: SunraChatCompletionParams = {
      model: 'google/gemini-2.5-flash',
      // Omit provider to use automatic routing.
      provider: { only: ['google-vertexai'] },
      messages: [
        {
          role: 'system',
          content: 'You are a helpful assistant.',
        },
        {
          role: 'user',
          content: 'Explain provider routing in two short sentences.',
        },
      ],
      stream: true,
    }
    const stream = await client.chat.completions.create(request)

    console.log(chalk.blue('\n' + '='.repeat(50)))
    console.log(chalk.blue('Streaming Response:'))
    console.log(chalk.blue('='.repeat(50)))

    let fullResponse = ''
    for await (const chunk of stream) {
      const content = chunk.choices[0]?.delta?.content
      if (content) {
        process.stdout.write(content)
        fullResponse += content
      }
    }

    console.log(chalk.blue('\n' + '='.repeat(50)))
    console.log(chalk.green(`Full Response: ${fullResponse}`))
    console.log(chalk.blue('='.repeat(50)))
  } catch (error) {
    const data = (error as any).response?.data
    // Check if this is an axios/fetch error with response data
    if (data) {
      console.error(chalk.red('Error:'), data)
    } else {
      console.error(chalk.red('Error:'), error)
    }
    process.exit(1)
  }
}

main().catch(console.error)

File: examples/demo-python/demo-llm.py
#!/usr/bin/env python3
"""
Demo script for LLM endpoint using OpenAI SDK directly.
This demonstrates how to use Sunra's LLM API with the standard OpenAI client.
"""

import os
from openai import OpenAI

def main():
    # Get API key from environment variable
    api_key = os.getenv('SUNRA_KEY')
    if not api_key:
        print("Error: SUNRA_KEY environment variable is not set")
        print("Get your API key from https://sunra.ai/dashboard/api-tokens")
        return

    # Initialize OpenAI client with Sunra endpoint
    client = OpenAI(
        api_key=api_key,
        base_url="https://api-llm.sunra.ai/v1"
    )

    try:
        print("Sending streaming chat completion request...")

        # Omit extra_body to use automatic provider routing.
        stream = client.chat.completions.create(
            model="google/gemini-2.5-flash",
            messages=[
                {
                  "role": "system",
                  "content": "You are a helpful assistant."
                },
                {
                  "role": "user",
                  "content": "Explain provider routing in two short sentences."
                }
            ],
            stream=True,
            extra_body={"provider": {"only": ["google-vertexai"]}}
        )

        print("\n" + "="*50)
        print("Streaming Response:")
        print("="*50)

        full_response = ""
        for chunk in stream:
            if chunk.choices[0].delta.content is not None:
                content = chunk.choices[0].delta.content
                print(content, end="", flush=True)
                full_response += content

        print("\n" + "="*50)
        print(f"Full Response: {full_response}")
        print("="*50)

    except Exception as error:
        print(f"Error: {error}")
        return

if __name__ == "__main__":
    main()
